modules/evm/src/lib.rs

```diff
@@ -837,29 +837,33 @@ impl<T: Config> EVMTrait for Module<T> {
 	type Balance = BalanceOf<T>;
 
 	fn execute(
-		source: EvmAddress,
-		target: EvmAddress,
+		context: InvokeContext,
 		input: Vec<u8>,
 		value: BalanceOf<T>,
 		gas_limit: u32,
 		storage_limit: u32,
-		config: Option<evm::Config>,
+		mode: ExecutionMode,
 	) -> Result<CallInfo, sp_runtime::DispatchError> {
+		let mut config = T::config().clone();
+		if let ExecutionMode::EstimateGas = mode {
+			config.estimate = true;
+		}
+
 		let info = Runner::<T>::call(
-			source,
-			target,
+			context.origin,
```

Comment:
should add new parameter to `Runner::<T>::call` to take both the sender & origin. sender should remain the same but vicinity.origin should be origin